File: examples/web-programs/web_review.mjs
import http from "node:http";
import https from "node:https";
const LIMIT = 262144;

function request(url, mode) {
  const u = new URL(url);
  if (!["http:", "https:"].includes(u.protocol) || u.username || u.password || u.hash) throw new Error("Invalid URL");
  return new Promise((resolve, reject) => {
    const headers = {"User-Agent":"Kali-Fieldbook-Lab/1.0", "Accept-Encoding":"identity"};
    if (mode === "cors") headers.Origin = "https://review.invalid";
    const req = (u.protocol === "https:" ? https : http).get(u, {headers, maxHeaderSize:16384}, res => {
      let size = 0;
      res.on("data", chunk => {size += chunk.length; if (size > LIMIT) req.destroy(new Error("Body too large"));});
      res.on("error", reject);
      res.on("end", () => {
        if (size > LIMIT) return;
        const values = new Map();
        for (let i=0; i<res.rawHeaders.length; i+=2) {
          const name = res.rawHeaders[i].toLowerCase();
          values.set(name, [...(values.get(name) ?? []), res.rawHeaders[i+1]]);
        }
        resolve({status:res.statusCode, values});
      });
    });
    const timer = setTimeout(() => req.destroy(new Error("Deadline")), 3000);
    req.on("close", () => clearTimeout(timer));
    req.on("error", reject);
  });
}
function analyze(mode, {status, values}) {
  if (mode === "headers") {
    const names = ["content-security-policy","strict-transport-security","x-content-type-options","referrer-policy","x-frame-options"];
    return {status, present:Object.fromEntries(names.map(n => [n,values.has(n)])), note:"Presence is not proof of correct policy; inspect application context."};
  }
  if (mode === "cookies") {
    const cookies = (values.get("set-cookie") ?? []).map(raw => {
      const parts = raw.split(";").map(x => x.trim());
      const separator = parts[0].indexOf("=");
      if (separator < 0) return {parse_error:"Missing cookie name/value separator"};
      const attributes = new Map(parts.slice(1).map(part => {
        const at = part.indexOf("=");
        return at < 0 ? [part.toLowerCase(), ""] : [part.slice(0,at).toLowerCase(), part.slice(at+1)];
      }));
      return {name:parts[0].slice(0,separator), secure:attributes.has("secure"), httponly:attributes.has("httponly"), samesite:attributes.get("samesite") ?? null, domain_attribute:attributes.has("domain")};
    });
    return {status,cookies,note:"Values omitted. Attribute summary is not a full browser cookie-policy validator."};
  }
  const allow_origin = values.get("access-control-allow-origin") ?? [];
  const allow_credentials = values.get("access-control-allow-credentials") ?? [];
  return {status,allow_origin,allow_credentials,review_reflection:allow_origin.length===1 && allow_origin[0]==="https://review.invalid" && allow_credentials.length===1 && allow_credentials[0]==="true",note:"One synthetic Origin probe; browser behavior, credentials and sensitive data must be assessed separately."};
}
const [mode,url] = process.argv.slice(2);
if (process.argv.length !== 4 || !["headers","cookies","cors"].includes(mode)) {
  console.error("Usage: node web_review.mjs headers|cookies|cors URL"); process.exitCode=2;
} else {
  try {console.log(JSON.stringify(analyze(mode,await request(url,mode))));}
  catch {console.error("Review failed: request or input error");process.exitCode=1;}
}

File: examples/web-programs/web_review.py
import argparse, http.client, json, ssl
from urllib.parse import urlsplit
LIMIT = 262144

def request(url, origin=None):
    u = urlsplit(url)
    if u.scheme not in ("http", "https") or not u.hostname or u.username or u.password or u.fragment:
        raise ValueError("Use an HTTP(S) URL without credentials or fragment")
    port = u.port or (443 if u.scheme == "https" else 80)
    cls = http.client.HTTPSConnection if u.scheme == "https" else http.client.HTTPConnection
    connection = cls(u.hostname, port, timeout=3)
    headers = {"User-Agent": "Kali-Fieldbook-Lab/1.0", "Accept-Encoding": "identity"}
    if origin is not None:
        headers["Origin"] = origin
    try:
        connection.request("GET", (u.path or "/") + ("?" + u.query if u.query else ""), headers=headers)
        response = connection.getresponse()
        body = response.read(LIMIT + 1)
        if len(body) > LIMIT:
            raise ValueError("Response exceeds 256 KiB")
        return response.status, response.getheaders()
    finally:
        connection.close()

def analyze(mode, status, headers):
    values = {}
    for name, value in headers:
        values.setdefault(name.lower(), []).append(value)
    if mode == "headers":
        names = ["content-security-policy", "strict-transport-security", "x-content-type-options", "referrer-policy", "x-frame-options"]
        return {"status": status, "present": {n: n in values for n in names}, "note": "Presence is not proof of correct policy; inspect application context."}
    if mode == "cookies":
        result = []
        for raw in values.get("set-cookie", []):
            parts = [part.strip() for part in raw.split(";")]
            if "=" not in parts[0]:
                result.append({"parse_error": "Missing cookie name/value separator"})
                continue
            name = parts[0].split("=", 1)[0]
            attributes = {}
            for part in parts[1:]:
                key, _, value = part.partition("=")
                attributes[key.lower()] = value
            result.append({"name": name, "secure": "secure" in attributes, "httponly": "httponly" in attributes, "samesite": attributes.get("samesite"), "domain_attribute": "domain" in attributes})
        return {"status": status, "cookies": result, "note": "Values omitted. Attribute summary is not a full browser cookie-policy validator."}
    acao = values.get("access-control-allow-origin", [])
    acac = values.get("access-control-allow-credentials", [])
    return {"status": status, "allow_origin": acao, "allow_credentials": acac, "review_reflection": acao == ["https://review.invalid"] and acac == ["true"], "note": "One synthetic Origin probe; browser behavior, credentials and sensitive data must be assessed separately."}

def main():
    p = argparse.ArgumentParser(description="One GET; no redirects; no supplied credentials. Use your approved URL.")
    p.add_argument("mode", choices=["headers", "cookies", "cors"])
    p.add_argument("url")
    a = p.parse_args()
    try:
        status, headers = request(a.url, "https://review.invalid" if a.mode == "cors" else None)
        print(json.dumps(analyze(a.mode, status, headers), sort_keys=True))
    except (ValueError, OSError, http.client.HTTPException) as error:
        p.exit(1, f"Review failed: {type(error).__name__}\n")

if __name__ == "__main__": main()
